.emacs.d/config.org

```diff
@@ -55,62 +55,392 @@
+
+  #+BEGIN_SRC emacs-lisp
+  (use-package which-key
+    :init
+    (which-key-mode))
+  #+END_SRC
+
+* Linters
+
+  The two main frameworks that allow Emacs to interface with external linters are
+  - [[https://www.gnu.org/software/emacs/manual/html_node/emacs/Flymake.html][flymake]]
+  - [[https://www.flycheck.org/en/latest/][flycheck]]
+  A detailed comparison is at https://www.flycheck.org/en/latest/user/flycheck-versus-flymake.html. We choose to use flycheck due to the sheer number of extensions written for it; pick your favorite language linter, and there’s probably a flycheck extension for it.
+
+  #+BEGIN_SRC emacs-lisp
+  (use-package flycheck)
+  #+END_SRC
+
+* Language and debug servers
+
+  [[https://microsoft.github.io/language-server-protocol][LSP]] is a protocol used to implement IDE-like features in an editor-agnostic manner. This package will give us a uniform way to add features such as semantic highlighting, auto-completion, jump-to-definition, and so on.
+
+  #+BEGIN_SRC emacs-lisp
+  (use-package lsp-mode
+    :commands lsp
+    :hook ((c-mode c++-mode python-mode sh-mode) . lsp)
+    :custom
+    (lsp-auto-guess-root t)
+    (lsp-log-io t)
+    (lsp-prefer-flymake nil))
+  #+END_SRC
+
+  This package enhances the user interface, providing pop-up windows with documentation and visual indicators.
+
+  #+BEGIN_SRC emacs-lisp
+  (use-package lsp-ui
+    :hook (lsp-mode . lsp-ui-mode)
+    :custom
+    (lsp-ui-doc-delay 3.0)
+    (lsp-ui-doc-include-signature t)
+    (lsp-ui-flycheck-enable t)
+    (lsp-ui-peek-always-show t))
+  #+END_SRC
+
+  Analogous to LSP is the [[https://microsoft.github.io/debug-adapter-protocol/][debug adapter protocol]]. We use [[https://github.com/emacs-lsp/lsp-treemacs][treemacs]] for IDE-like display of errors.
+
+  #+BEGIN_SRC emacs-lisp
+  (use-package dap-mode
+    :after lsp-mode
+    :config
+    (dap-mode t)
+    (dap-ui-mode t))
+
+  (use-package treemacs
+    :bind (:map global-map
+                ("C-x t t" . treemacs)
+                ("C-x t 1" . treemacs-select-window))
+    :custom
+    (treemacs-resize-icons 15))
+
+  (use-package lsp-treemacs
+    :init (lsp-treemacs-sync-mode 1)
+
+    :bind (:map java-mode-map
+                ("C-x e l" . lsp-treemacs-errors-list)
+                ("C-x s l" . lsp-treemacs-symbols)))
+  #+END_SRC
+
+* Code completion
+
+  Company is the primary package that is used for code completion, it follows a frontend/backend system. The package =company= is the frontend, it will query a certain backend based on what code you are editing, such as one provided by an active language server.
+
+  #+BEGIN_SRC emacs-lisp
+  (use-package company
+    :hook (after-init . global-company-mode)
+    :custom
+    (company-tooltip-align-annotations t))
+
+  (use-package company-lsp
+    :after (company lsp-mode)
+    :custom
+    (company-lsp-cache-candidates t))
+  #+END_SRC
+
+* Languages
+
+** HTML / CSS / HTML Templates
+
```

Comment:
Might be helpful to throw in this [cheatsheet](https://docs.emmet.io/cheatsheet-a5.pdf). And perhaps a mini [tutorial video](https://www.youtube.com/watch?v=V8vizNQKtx0)